lib/Backend/GlobOpt.cpp

```diff
@@ -18456,34 +18459,68 @@ GlobOpt::OptHoistInvariant(
     Value *dstVal,
     Value *const src1Val,
     bool isNotTypeSpecConv,
-    bool lossy)
+    bool lossy,
+    IR::BailOutKind bailoutKind)
 {
     BasicBlock *landingPad = loop->landingPad;
     IR::RegOpnd *dst = instr->GetDst() ? instr->GetDst()->AsRegOpnd() : nullptr;
     if(dst)
     {
-        switch(instr->m_opcode)
-        {
-            case Js::OpCode::CmEq_I4:
-            case Js::OpCode::CmNeq_I4:
-            case Js::OpCode::CmLt_I4:
-            case Js::OpCode::CmLe_I4:
-            case Js::OpCode::CmGt_I4:
-            case Js::OpCode::CmGe_I4:
-            case Js::OpCode::CmUnLt_I4:
-            case Js::OpCode::CmUnLe_I4:
-            case Js::OpCode::CmUnGt_I4:
-            case Js::OpCode::CmUnGe_I4:
-                // These operations are a special case. They generate a lossy int value, and the var sym is initialized using
-                // Conv_Bool. A sym cannot be live only as a lossy int sym, the var needs to be live as well since the lossy int
-                // sym cannot be used to convert to var. We don't know however, whether the Conv_Bool will be hoisted. The idea
-                // currently is that the sym is only used on the path in which it is initialized inside the loop. So, don't
-                // hoist any liveness info for the dst.
-                if (!this->GetIsAsmJSFunc())
-                {
-                    lossy = true;
+        switch (instr->m_opcode)
+        {
+        case Js::OpCode::CmEq_I4:
+        case Js::OpCode::CmNeq_I4:
+        case Js::OpCode::CmLt_I4:
+        case Js::OpCode::CmLe_I4:
+        case Js::OpCode::CmGt_I4:
+        case Js::OpCode::CmGe_I4:
+        case Js::OpCode::CmUnLt_I4:
+        case Js::OpCode::CmUnLe_I4:
+        case Js::OpCode::CmUnGt_I4:
+        case Js::OpCode::CmUnGe_I4:
+            // These operations are a special case. They generate a lossy int value, and the var sym is initialized using
+            // Conv_Bool. A sym cannot be live only as a lossy int sym, the var needs to be live as well since the lossy int
+            // sym cannot be used to convert to var. We don't know however, whether the Conv_Bool will be hoisted. The idea
+            // currently is that the sym is only used on the path in which it is initialized inside the loop. So, don't
+            // hoist any liveness info for the dst.
+            if (!this->GetIsAsmJSFunc())
+            {
+                lossy = true;
+            }
+            break;
+
+        case Js::OpCode::FromVar:
+        {
+            StackSym* src1StackSym = IR::RegOpnd::TryGetStackSym(instr->GetSrc1());
+
+            // We may be hoisting FromVar from a region where it didn't need a bailout (src1 had a definite value type) to a region
+            // where it would. In such cases, the FromVar needs a bailout based on the value type of src1 in its new position.
+            if (src1StackSym && !instr->HasBailOutInfo() && bailoutKind != IR::BailOutInvalid)
```

Comment:
Probably should have an assert that if it has bailout, it is one of the expected ones.